### PR TITLE
feat(router): added incoming refund webhooks flow

### PR DESCRIPTION
**Pull request by saiharsha-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Added support for incoming refund webhooks. 

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Manually tested
Webhook Request
image

Outgoing webhook to merchant
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/adyen.rs [hunk 1]
```diff
@@ -735,6 +735,9 @@ impl api::IncomingWebhook for Adyen {
 
         Ok(match notif.event_code.as_str() {
             "AUTHORISATION" => api::IncomingWebhookEvent::PaymentIntentSuccess,
+            "REFUND" => api::IncomingWebhookEvent::RefundSuccess,
+            "CANCEL_OR_REFUND" => api::IncomingWebhookEvent::RefundSuccess,
+            "REFUND_FAILED" => api::IncomingWebhookEvent::RefundFailure,
```
Comment (SanchithHegde) [inline crates/router/src/connector/adyen.rs:740]: If these event codes have been documented on their docs, please convert this to an enum instead.

File: crates/router/src/core/webhooks.rs [hunk 2]
```diff
@@ -90,6 +90,99 @@ async fn payments_incoming_webhook_flow(
...
+            .await
+            .change_context(errors::WebhooksFlowError::RefundsCoreFailed)
+            .attach_printable_lazy(|| {
+                format!(
+                    "Failed while updating refund: refund_id: {}",
+                    refund.refund_id
+                )
+            })?
+    };
+    let refund_id = updated_refund.refund_id.clone();
```
Comment (SanchithHegde) [inline crates/router/src/core/webhooks.rs:162]: The refund ID for the updated refund must be the same as before. This reassignment is not necessary.

**Review by SanchithHegde — CHANGES_REQUESTED** [review]
Other than that, looks good to me!

File: crates/router/src/core/webhooks.rs (hunk 2 above)
Comment (saiharsha-juspay) [reply crates/router/src/core/webhooks.rs:162]: Yes, will remove that!

File: migrations/2022-09-29-084920_create_initial_tables/up.sql [hunk 3]
```diff
@@ -143,11 +143,11 @@ CREATE TYPE "Currency" AS ENUM (
...
 );
 
-CREATE TYPE "EventClass" AS ENUM ('payments');
+CREATE TYPE "EventClass" AS ENUM ('payments', 'refunds');
 
-CREATE TYPE "EventObjectType" AS ENUM ('payment_details');
+CREATE TYPE "EventObjectType" AS ENUM ('payment_details', 'refund_details');
 
-CREATE TYPE "EventType" AS ENUM ('payment_succeeded');
+CREATE TYPE "EventType" AS ENUM ('payment_succeeded', 'refund_succeeded', 'refund_failed');
```
[the remark this answers is not in the document]
Comment (saiharsha-juspay) [reply migrations/2022-09-29-084920_create_initial_tables/up.sql:150]: Okay, adding the migrations

File: crates/router/src/core/webhooks.rs [hunk 4]
```diff
@@ -90,6 +90,99 @@ async fn payments_incoming_webhook_flow(
     Ok(())
 }
 
+fn get_refund_status_from_event_type(
+    event_type: api_models::webhooks::IncomingWebhookEvent,
+) -> CustomResult<storage_models::enums::RefundStatus, errors::WebhooksFlowError> {
```
[the remark this answers is not in the document]
Comment (saiharsha-juspay) [reply crates/router/src/core/webhooks.rs:95]: Yes, will change it to TryFrom

File: crates/router/src/connector/adyen.rs (hunk 1 above)
Comment (saiharsha-juspay) [reply crates/router/src/connector/adyen.rs:740]: okay

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -989,14 +989,26 @@ pub struct AdyenAmountWH {
...
+pub enum WebhookEventCode {
+    #[serde(rename = "AUTHORISATION")]
+    Authorisation,
+    #[serde(rename = "REFUND")]
+    Refund,
+    #[serde(rename = "CANCEL_OR_REFUND")]
+    CancelOrRefund,
+    #[serde(rename = "REFUND_FAILED")]
+    RefundFailed,
+}
```
Comment (SanchithHegde) [inline crates/router/src/connector/adyen/transformers.rs:1002]:
```suggestion
#[derive(Debug, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")
pub enum WebhookEventCode {
    Authorisation,
    Refund,
    CancelOrRefund,
    RefundFailed,
}
```

File: migrations/2022-09-29-084920_create_initial_tables/up.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanchithHegde) [reply migrations/2022-09-29-084920_create_initial_tables/up.sql:150]: Please revert these changes

File: crates/router/src/connector/adyen/transformers.rs (hunk 5 above)
Comment (saiharsha-juspay) [reply crates/router/src/connector/adyen/transformers.rs:1002]: Changed!

File: migrations/2022-09-29-084920_create_initial_tables/up.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (saiharsha-juspay) [reply migrations/2022-09-29-084920_create_initial_tables/up.sql:150]: okay

**Review by SanchithHegde — APPROVED** [review]
Apart from the pg_enum table being modified in the `down.sql` file, looks good to me!

File: crates/router/src/connector/adyen/transformers.rs [hunk 6]
```diff
@@ -989,14 +989,23 @@ pub struct AdyenAmountWH {
...
 
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum WebhookEventCode {
+    Authorisation,
+    Refund,
+    CancelOrRefund,
+    RefundFailed,
+}
+
```
Comment (vspecky) [inline crates/router/src/connector/adyen/transformers.rs:1000]: Derive `strum::Display` and annotate with `#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]` to get a display implementation.

File: crates/router/src/connector/adyen.rs [hunk 7]
```diff
@@ -688,7 +688,7 @@ impl api::IncomingWebhook for Adyen {
             .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
 
         let message = format!(
-            "{}:{}:{}:{}:{}:{}:{}:{}",
+            "{}:{}:{}:{}:{}:{}:{:?}:{}",
```
Comment (vspecky) [inline crates/router/src/connector/adyen.rs:691]: Use a display implementation for `event_code`

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/adyen/transformers.rs (hunk 6 above)
Comment (saiharsha-juspay) [reply crates/router/src/connector/adyen/transformers.rs]: Done

File: crates/router/src/connector/adyen.rs (hunk 7 above)
Comment (saiharsha-juspay) [reply crates/router/src/connector/adyen.rs]: Okay

**Review by vspecky — APPROVED** [review]
LGTM

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/connector/adyen.rs [hunk 8]
```diff
@@ -733,9 +733,13 @@ impl api::IncomingWebhook for Adyen {
...
-        Ok(match notif.event_code.as_str() {
-            "AUTHORISATION" => api::IncomingWebhookEvent::PaymentIntentSuccess,
-            _ => Err(errors::ConnectorError::WebhookEventTypeNotFound).into_report()?,
+        Ok(match notif.event_code {
+            adyen::WebhookEventCode::Authorisation => {
+                api::IncomingWebhookEvent::PaymentIntentSuccess
+            }
+            adyen::WebhookEventCode::Refund => api::IncomingWebhookEvent::RefundSuccess,
+            adyen::WebhookEventCode::CancelOrRefund => api::IncomingWebhookEvent::RefundSuccess,
+            adyen::WebhookEventCode::RefundFailed => api::IncomingWebhookEvent::RefundFailure,
```
Comment (dracarys18) [inline crates/router/src/connector/adyen.rs:742]: This can be converted to `From` implementation?

**Review by dracarys18 — CHANGES_REQUESTED** [review]

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/connector/adyen.rs (hunk 8 above)
Comment (saiharsha-juspay) [reply crates/router/src/connector/adyen.rs]: Done

**Review by dracarys18 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Review by vspecky — APPROVED** [review]
LGTM

**Pull request merged**